### PR TITLE
add reduce=True arg to HingeEmbeddingLoss

### PR DESCRIPTION
**Pull request by li-roy**
As per #264. When reduce is False, HingeEmbeddingLoss outputs a loss per sample in minibatch. When reduce is True (default), the current behavior is kept.

Test Plan
test/run_test.sh
Added unit tests for the reduce=False case. Added a reference function.

### CODE REVIEW

File: test/common_nn.py [hunk 1]
```diff
@@ -364,12 +364,29 @@ def multilabelmarginloss_reference(input, target, size_average=True, reduce=True
     return output / dim
 
 
+def hingeembeddingloss_reference(input, target, margin=1.0, size_average=True, reduce=True):
+    # needed for legacy tests
+    if not isinstance(input, Variable):
```
Comment (zou3519) [inline test/common_nn.py]: Why is this needed for the legacy tests?

**Review by zou3519 — COMMENTED** [review]
Looks good to me. Just a quick comment about the legacy test.

File: test/common_nn.py (hunk 1 above)
Comment (li-roy) [reply test/common_nn.py]: torch.where function only works for Variables. I also tried using the Tensor.map_, which worked for the legacy test but broke test_nn. Let me know if you think there's a better way to do this though.

**Review by zou3519 — APPROVED** [review]
LGTM
